### PR TITLE
IKeyboard / Keyboard refactoring

### PR DESCRIPTION
**Pull request by dpisanu**
Cleaning up the Keyboard components

BareMetalKeyboard

This contains the DLLImports from user32
Basic Keyboard interaction and input validation
IBaseKeyboard / BaseKeyboard

Moved alot of the Keyboard Functions and Properties from Keyboard to
BaseKEyboard
BaseKeyboard handles anything not relevant to an IActionListener
IKeyboard / Keyboard

Handles anything related to an IActionListener
AttachedKeyboard

Implements IBaseKeyboard
Handles focusing on an UI Item Container before doing any keyboard
interaction

### CODE REVIEW

File: src/TestStack.White.sln [hunk 1]
```diff
@@ -1,7 +1,7 @@
 ﻿
 Microsoft Visual Studio Solution File, Format Version 12.00
 # Visual Studio 2013
-VisualStudioVersion = 12.0.40629.0
+VisualStudioVersion = 12.0.31101.0
```
Comment (Roemer) [inline src/TestStack.White.sln]: Don't change / commit this.

File: src/TestStack.White/InputDevices/BareMetaKeyboard.cs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+using System.Runtime.InteropServices;
+using TestStack.White.WindowsAPI;
+
+namespace TestStack.White.InputDevices
+{
+    //BUG: KeysConverter
+    /// <summary>
+    /// Represents Keyboard attachment to the machine.
+    /// </summary>
+    public static class BareMetaKeyboard
```
Comment (Roemer) [inline src/TestStack.White/InputDevices/BareMetaKeyboard.cs]: Missing "l" in the name / filename.

File: src/TestStack.White/InputDevices/BaseKeyboard.cs [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+namespace TestStack.White.InputDevices
+{
+    // BUG: KeysConverter
+    /// <summary>
+    /// Represents Keyboard attachment to the machine.
+    /// </summary>
+    public abstract class BaseKeyboard : IBaseKeyboard
+    {
+        #region Fields
```
Comment (JakeGinnivan) [inline src/TestStack.White/InputDevices/BaseKeyboard.cs]: No regions please

File: src/TestStack.White/InputDevices/IKeyboard.cs [hunk 4]
```diff
@@ -1,13 +1,42 @@
+using TestStack.White.UIItems.Actions;
 using TestStack.White.WindowsAPI;
 
 namespace TestStack.White.InputDevices
 {
-    public interface IKeyboard
+    public interface IKeyboard : IBaseKeyboard
```
Comment (JakeGinnivan) [inline src/TestStack.White/InputDevices/IKeyboard.cs]: What is the reason for the multiple interfaces?

**Comment by JakeGinnivan** [comment]
What are the benefits of this refactoring?

This introduces another 3 types to the keyboard infrastructure for what I can see is no benefit to the end users of white?

**Comment by dpisanu** [comment]
I am gonna try to answer this for this change and the Mouse Change.

Keyboard and Mouse have their respective Interfaces and static methods that do some of the bare work on the DLL Imports.

Both Keyboard and Mouse contain more public functions and properties then their Interfaces publish.
So the idea was to get them in synch and hence worth only pass the Interface around if required and not the concrete implementing Class.

Since AttachedKeyboard and AttachedMouse also implement the IKeyboard and IMouse Iinterface we come into the dilemma that the Interfaces hold some functions that are not required in the AttachedKeyboard or AttachedMouse.

Specifically the ones centered on the IActionListener.
The AttachedXYZ versions hold their own IActionListener  they want to do Actions on.
So the AttachedXYZ just require a Subset of the interface functions and wrap the real Keyboard or Mouse Instance inside doing some pre or post steps.

1# Split the Import code out into another static class
2# Create a Subset that is valid for Keyboard and AttachedKeyboard (Interface and Abstract Class)
3# Manage only the Functions related to IActionLIstener in the corresponding Interface for IKeyboard and IMouse.
4# The AttachedXYZ implement the Interface from Step 2#

From my opinion this makes it easier to maintain.
- (Bare Metal Work, Base Functionality, concrete implementations)
- Lays the ground work for only the Interfaces to be referenced by consumers and not the concrete Class Implementation. (Contracts)
- Change should not break any functionality and customer side implementations.

**Comment by dpisanu** [comment]
Took out the regions by request.
Fixed Typo.
Fixed version change in Solution. (my bad)

**Comment by JakeGinnivan** [comment]
Ok, 

@Roemer and @dpisanu, I like that we are trying to improve the codebase. How about we start targeting these refactorings to the UIA Com Wrapper branch?

From my point of view, there is a chunk of work that needs to be done on that branch to make the control dictionary go away (the design is broken for the v3 UIA libraries). But White really has no future not switching to those new UIA libraries. 

Once that branch is functional, we can cut v1 with any public API changes we want and White will be in a really good place going forward. My preference is to target the hard problems first rather than refactorings which are more academic in nature (which do not impact the consumers of the library that much).

Maybe we should have a skype call some time to discuss all this?

**Comment by Roemer** [comment]
I'm totally with you that we should focus on the UIACom branch. I see the "refactoring" tasks as learning the current code base. Any why not push those things when we edit them anyway. We could think about making those refactorings in the UIACom branch only, which possibly means that, at some point, we will replace the current master with the UIACom branch instead of merge.
We should try gathering a few "core" people who will work on the UIACom branch and have a possible Skype kick-off.

**Comment by dpisanu** [comment]
Most of my changes are basecally on the code base that is being shipped and that i use productively.
It's things that come up on a day to day use and I see as an improvement to the working code base that current users would benefit from.

If we decive that these kind of changes should primarily go into the UAICom branch, I'll be happy to comply and do it there.

Do you see current version and UIACom living in parallel in the future or will UIACom completely supersede the current version?
What changes are worth having in both branches and which not?

**Comment by JakeGinnivan** [comment]
The UIAComWrapper branch will supersede the current master, but it is not as stable atm. 

The more changes on this branch, the harder the merge will be and issues will be introduced. So I recon just work on the UIA Com Wrapper branch, it gives virtualisation support and other newer automation patterns. Also perf improvements and such

**Comment by dpisanu** [comment]
So is this change ok to go on master? I would then create a separate Changeset for the UIACom Branch

**Comment by Roemer** [comment]
We should probably only do bugfixing from now on in the master branch and really focus on the uiacom branch. So I think our three newest PRs (mouse, keyboard, config) should go directly to the uiacom branch.

**Comment by dpisanu** [comment]
do you want to take care of putting them into the UIACom Branch or should I make a new PR for that branch ?

**Comment by Roemer** [comment]
You can create new pull requests for the mouse/keyboard changes. I will backmerge the current master to uiacom soon.

**Comment by dpisanu** [comment]
Please close this PR

**Pull request closed without merging**